### ISSUE TITLE
Other Media files and admin

### ISSUE THREAD

**Issue by susanodd** [Signbank/Global-signbank#735]
Improvements to the interface to make management of other media easier for the user.
Followup to #440 

- Gloss/Morpheme Display checks existence of Other Media files on server, user feedback
- Gloss/Morpheme Edit checks filename and file type during Upload, user feedback

**Comment by susanodd** [comment]
This has been pushed to github.

**Comment by ocrasborn** [comment]
Hey @susanodd , I just noticed that for the CSL_Shanghai dataset, all 'main' movies have disappeared, but that these very movies are now available under Other Media. Is this related to this issue? Or something else you guys are working on?

**Comment by susanodd** [comment]
I'm not sure. @vanlummelhuizen has just merged the branch. I don't know if it was live when you posted that or not?

The other media is stored under the gloss id (folder) under (protected_media) othermedia.
The videos are stored under (protected_media) glossvideo under the dataset acronym (folder).
Nothing was moved.

It could be that broken paths are now working. There may have been a path prefix missing?
(I'll have to search to figure it out.)
There is also a folder called import_videos. But that is connected to a batch import. @Woseseltops was working on that. #717

**Comment by vanlummelhuizen** [comment]
> I'm not sure. @vanlummelhuizen has just merged the branch. I don't know if it was live when you posted that or not?

@susanodd I merged your work and put it live after @ocrasborn 's comment.

**Comment by ocrasborn** [comment]
@susanodd @vanlummelhuizen : I looked on the server under glossvideo/, and see that in the "old" (?) CSL_Shanghai/, there's only empty subdirectories, while there's a new folder CSL_Shangh/ that contains all the videos. I could simply clean things up there manually, but I'd best leave that to you, and more importantly: how could this have come about?!

**Comment by susanodd** [comment]
Yes, it's best to check whether there are paths to the old locations inside glosses.
Was the dataset renamed?

**Comment by susanodd** [comment]
I'm looking at the GlossVideo objects (on my local server) to see what the paths look like.
All of the paths with "CSL" in them refer to folder glossvideo/CSL_Shanghai/ ....
I tried looking at one of these glosses (6675), but there are no "Other Media" files for it (nor on the live server).
Have you got a gloss ID for where this happens?

**Comment by susanodd** [comment]
Would it fix it to just rename CSL_Shangh ? It should be CSL_Shanghai 

Maybe #717 did something?

https://github.com/Signbank/Global-signbank/blob/511bc02da4e670ade6f932bd95b3aa08e8a0cdd2/signbank/dictionary/views.py#L503

https://github.com/Signbank/Global-signbank/blob/511bc02da4e670ade6f932bd95b3aa08e8a0cdd2/signbank/dictionary/views.py#L522-L525


There was a similar import routine for other media, but I commented out the url because it wanted to put the files in a wrong location and the code was obsolete. It wanted to put files in static, and it was looking for a file index.csv.

https://github.com/Signbank/Global-signbank/blob/511bc02da4e670ade6f932bd95b3aa08e8a0cdd2/signbank/dictionary/views.py#L616

**Comment by ocrasborn** [comment]
I'm happy just renaming the folder, but I'd also want to find out how it came about. If you know enough to understand what might have gone on, could you rename the folder (and remove the other one) for me, @susanodd ?

**Comment by susanodd** [comment]
Okay, I downloaded the most recent database to check there. My previous local database was from 31 march. The new one is from yesterday. The GlossVideo path names are all with CSL_Shanhai.

I'm logged onto the server now.
There are indeed two folders under glossvideo.
The one CSL_Shanghai is from 2019 created by @vanlummelhuizen the other CSL_Shangh is from 2020 created by @Woseseltops .
@ocrasborn Is it possible you didn't try to access videos for a while?
(Probably not.)

Looking further, indeed, both folders have the same number of two-letter subfolders, created resp. by the two cohorts.
All created on 21 august 2020.
And indeed, the folders under CSL_Shanghai are empty, but those under CSL_Shangh are not.
It kind of looks like a typo or something. Esp. that it's two users, rather than the signbank website.
Perhaps somebody did a command line command? @vanlummelhuizen @Woseseltops 
I don't think code caused this, it's too clean.

**Comment by susanodd** [comment]
I'm okay with just renaming. I added the lengthy discussion above in case @vanlummelhuizen and @Woseseltops want to comment, or indeed remember doing something.

**Comment by ocrasborn** [comment]
OK! Go ahead then, @susanodd

**Comment by susanodd** [comment]
I found one directory in CSL_Shanghai that actually has a video. LAUNDRY-MACHINE-A
(Added in January.)
I suspect I need to do it all by hand and remove the old empty ones, and move this one and any others, etc.
Check.

**Comment by ocrasborn** [comment]
I found two yesterday, but they seem to be linked to glosses we deleted, feel free to ignore and thus delete those

**Comment by susanodd** [comment]
Okay, I found a total of 7 videos and moved these to the "other" folders.
Now it's all repaired.

I noticed that the videos don't show up in the gloss search page column, but they are all there if you look at the glosses.
E.g., searching for ^LA
there are quite a few, 17 glosses. but only a preview video of LAUNDRY-MACHINE-A shows up in the gloss list.

**Comment by vanlummelhuizen** [comment]
I tested which video files in the list of GlossVideos of CSL_Shanghai did not exist:

```
>>> for v in GlossVideo.objects.filter(gloss__lemma__dataset__acronym="CSL_Shanghai"):
...     if not os.path.exists('/var/www/signbank/live/writable/'+str(v.videofile)):
...         print(v.videofile)
... 
glossvideo/CSL_Shanghai/AC/ACCUSE-5149.mp4
glossvideo/CSL_Shanghai/AB/ABLE-BODIED-PEOPLE-B-6691.mp4
glossvideo/CSL_Shanghai/BA/BAOZI-B-6739.mp4
glossvideo/CSL_Shanghai/BE/BEARD-A-6746.mp4
glossvideo/CSL_Shanghai/BE/BEAT-B-6747.mp4
glossvideo/CSL_Shanghai/BE/BEGGAR-A-6749.mp4
glossvideo/CSL_Shanghai/BE/BEGINNING-OF-TERM-B-6751.mp4
glossvideo/CSL_Shanghai/BI/BIKE-B-6754.mp4
glossvideo/CSL_Shanghai/BL/BLUE-B-6757.mp4
glossvideo/CSL_Shanghai/BL/BLUNT-B-6758.mp4
glossvideo/CSL_Shanghai/BR/BRAKE-B-6772.mp4
glossvideo/CSL_Shanghai/CA/CAKE-B-6783.mp4
glossvideo/CSL_Shanghai/CA/CAMERA-B-6784.mp4
glossvideo/CSL_Shanghai/CA/CARD-B-6787.mp4
glossvideo/CSL_Shanghai/CH/CHRISTMAS-A-6806.mp4
glossvideo/CSL_Shanghai/CL/CLAW-A-6816.mp4
glossvideo/CSL_Shanghai/CO/CONFLICT-B-6838.mp4
glossvideo/CSL_Shanghai/CO/COOL-B-6843.mp4
glossvideo/CSL_Shanghai/DA/DAYTIME-B-6859.mp4
glossvideo/CSL_Shanghai/DR/DRIVER-B-6886.mp4
glossvideo/CSL_Shanghai/FA/FATHER-IN-LAW-C-6924.mp4
glossvideo/CSL_Shanghai/FI/FIRST-B-6933.mp4
glossvideo/CSL_Shanghai/PE/PEOPLE-WITH-PHYSICAL-DISABILITY-5526.mp4
glossvideo/CSL_Shanghai/LA/LAUNDRY-MACHINE-A-6153
```
The last one seems to have an easy fix: append .mp4 to the end of the file name in the database. The others, I don't know.

**Comment by susanodd** [comment]
I already appended the last one with mp4 !

**Comment by susanodd** [comment]
I just removed by hand a bunch of two-letter empty folders in the glossvideo folder. (i.e., not inside a dataset)

**Comment by vanlummelhuizen** [comment]
> I already appended the last one with mp4 !

On the live server? I don't that (yet?).

**Comment by susanodd** [comment]
These are the ones I moved by hand:
`./WH:
total 10236
-rw-rw----+ 1 mhulsbosch wwwsignbank 10478609 jun 27  2018 WHO-A-5391.mp4


./TR:
total 620
-rw-rw----+ 1 wstoop wwwsignbank 632638 mei 24  2018 TRANSLATE-B-5166.mp4


./SO:
total 1084
-rw-rw----+ 1 wstoop wwwsignbank 1107338 mei 24  2018 SOCCER-B-5008.mp4

./PE:
total 568
-rw-rw----+ 1 wstoop wwwsignbank 579568 mei 24  2018 PERSON-B-5265.mp4


./NO:
total 488
-rw-rw----+ 1 wstoop wwwsignbank 478667 mei 24  2018 NOT-B-5350.mp4
-rw-rw----+ 1 wstoop wwwsignbank  18789 mei 24  2018 NOT-B-5350_small.mp4

./LA:
total 212
-rw-rw----+ 1 wapsignbank wwwsignbank 215617 jan 26 15:37 LAUNDRY-MACHINE-A-6153`

**Comment by susanodd** [comment]
They're mixed in with the other files now, CSL_Shangh moved / renamed to CSL_Shanghai
With the last one now has .mp4

**Comment by susanodd** [comment]
There is something weird with the search results on the Gloss Detail page.
When you keep selecting another gloss from the search results bar, it keeps getting concatenated with a duplicate url.
`https://signbank.cls.ru.nl/dictionary/gloss/6284/dictionary/gloss/dictionary/gloss/dictionary/gloss/6195`

**Comment by susanodd** [comment]
That is actually a problem. It keeps showing the "first" gloss rather than the "next" gloss which has been concatenated at the end.

**Comment by susanodd** [comment]
https://signbank.cls.ru.nl/dictionary/gloss/5860/

https://signbank.cls.ru.nl/dictionary/gloss/5919

(That's what they should be. If you search on ^LA there are lots of glosses with videos.)

**Comment by susanodd** [comment]
Oh dear, same thing for glossimage!

**Comment by susanodd** [comment]
Okay, I also moved images!

**Comment by susanodd** [comment]
All of the spurious 2-char empty folders have been removed from glossvideo and glossimage.

One thing about the Dataset folders, Kata Kolok has a space in the filename. That seems wrong. Shouldn't it be an underscore?

**Comment by vanlummelhuizen** [comment]
> There is something weird with the search results on the Gloss Detail page.
> When you keep selecting another gloss from the search results bar, it keeps getting concatenated with a duplicate url.
> `https://signbank.cls.ru.nl/dictionary/gloss/6284/dictionary/gloss/dictionary/gloss/dictionary/gloss/6195`

This could be solved with the following patch:

```
diff --git a/signbank/dictionary/templates/dictionary/search_result_bar.html b/signbank/dictionary/templates/dictionary/search_result_bar.html
index f435e2d4..a676f76c 100755
--- a/signbank/dictionary/templates/dictionary/search_result_bar.html
+++ b/signbank/dictionary/templates/dictionary/search_result_bar.html
@@ -11,7 +11,7 @@
                 a.setAttribute("class", "search_result");
                 a.style["float"] = "none";
                 var type_of_data = json_data[key].href_type;
-                a.href = "{{PREFIX_URL}}dictionary/" + type_of_data + "/" + json_data[key].id;
+                a.href = "/{{PREFIX_URL}}dictionary/" + type_of_data + "/" + json_data[key].id;^M
                 a.id = json_data[key].id;
                 var data_label = json_data[key].data_label;
                 if (data_label) {
```

**Comment by ocrasborn** [comment]
The present version that's live presents me with an error when clicking the upload button that I haven't seen before:
>ImportError at /dictionary/update/othermedia/
No module named 'magic'

**Comment by susanodd** [comment]
That would be the Python library magic.
sorry I hadn't realised that wasn't part of the system modules.
It needs to be installed on the server Python.
It checks filename extensions.

**Comment by susanodd** [comment]
@Woseseltops or @vanlummelhuizen can you install this module on the server?

python-magic==0.4.22

**Comment by susanodd** [comment]
Bug fixes pushed to branch `lemmas_minimalpairs_733_743`

**Comment by vanlummelhuizen** [comment]
> @Woseseltops or @vanlummelhuizen can you install this module on the server?
> 
> python-magic==0.4.22

@susanodd @ocrasborn I just installed python-magic and restarted the Signbank.

**Comment by susanodd** [comment]
Great!

**Comment by susanodd** [comment]
@Woseseltops @ocrasborn @vanlummelhuizen everybody okay with this issue?

**Comment by Woseseltops** [comment]
I am reviewing this code, and if I understand correctly, Python magic is used to verify that for example an uploaded .mp4 file really is a video. Out of curiosity, @susanodd , was there some kind of problem or use case where file types did not match their extensions?

**Comment by susanodd** [comment]
It's used to determine the extension. There is an additional check when starting to copy the temp file, that the first block is checked also that it conforms and matches what magic determined.

I have a bunch of stuff on my computer that I kept testing it out on. So that's also why it checks the extension 'm4v' and changes it to 'mp4'. I have a couple of files I once downloaded (saved) from Signbank for the R to L languages and their filenames were constructed backwards. So sometimes it seemed like it was having trouble parsing those names.
So I wanted to make sure it works whatever, more robust.
It's possible to simply rename another file by putting an extension on it. (I tried this with renaming a csv file to be mp4.)

On some of the sites where this is discussed, they wrote about checking the actual type stored in the file to make sure it's what the extension is.
So basically long-winded checks to (help) make sure people aren't uploading weird stuff.
